### ISSUE TITLE
cyrus 3.2 sieve vacation dies with 'free(): invalid pointer' when subject is given

### ISSUE THREAD

**Issue by slangenhorst** [cyrusimap/cyrus-imapd#3060]
Hi,

another problem: if a user sets a subject in his vacation like
`vacation :days 7 :addresses ["foo@bar.com"] :subject "Wont come back" "Vacation so much nicer"`
lmtpd dies with "free(): invalid pointer" _after_ working through the sieve/lmtpd process. So with a sieve redirect you would get multiple copies, since the deliviering MTA just gets a killed connection after DATA and so on.

Problem seems the use of cmd.u.v.subject
With a given subject, the string never gets copied if "requires & BFE_VARIABLES" is false, so when it is free'd in sieve/message.c:592 it fails.
If no subject is given the auto-generated is buf_release'd and the free works fine.

Patch:
`--- a/sieve/bc_eval.c`
`+++ b/sieve/bc_eval.c`
`@@ -2124,6 +2124,8 @@`
`                 }`
`  `
`                 subject = buf_release(&buf);`
`+            } else {`
`+               subject = xstrdup(subject);`
`             }`
`  `
`             if (from) {`

**Comment by elliefm** [comment]
Have confirmed the crash with a new Cassandane test (only on my branch so far) -- affects both 3.2 and master: https://github.com/cyrusimap/cassandane/compare/master...elliefm:3060-sieve-subject-bad-free

The proposed patch prevents the crash, but I think it might introduce a memory leak when the script requires the "variables" extension?  Observe that the subject will be first `xstrdupnull()`'d when it is parsed for variables in the `(requires & BFE_VARIABLES)` block, and then again in the new `else` block.  The latter allocation will be freed eventually, but the first will be leaked?

Yep:

```
=====> Instance::_check_valgrind_logs[1409] Valgrind errors from file /dev/shm/cass/0725390101/vglogs/lmtpd.16833
=====> Instance::_check_valgrind_logs[1409] ==16833== 4 bytes in 1 blocks are definitely lost in loss record 2 of 138
```

It's getting late here, and I'm probably too tired to figure out a better patch right now... @ksmurchison, wanna have a look?  Otherwise I'll look at this further next week

**Comment by ksmurchison** [comment]
Fixed on master with 
https://github.com/cyrusimap/cyrus-imapd/commit/589dfbaa93da4f756774f86830d37b74db4ba419

**Comment by elliefm** [comment]
Thanks @ksmurchison, fix is now on the 3.2 branch as well, and will be included in 3.2.2
